Use normalized package names where suitable

DragaDoncila opened the pull request:
Closes #48 

This PR updates npe2api to use the **normalized** plugin name as the plugin key, and adds a `name` key to dictionaries holding the original, non-normalized name. Additionally, it ensures that all individual plugin files in the `public/pypi`, `public/conda` and `public/manifest` directories are named by their normalized name. 

To simplify retrieval of the manifests via the API, we normalize the name in the dynamic routing handler, so that the api endpoint for the manifest still uses the original name.

DragaDoncila commented:
I've left this PR as draft for now to get some feedback from you @aganders3 as to whether this is what you were imagining.

One thing is I think we'll need to **clear** the `public/manifest/` folder and re-write all the manifests in order for the slug to work and to make sure we've deduplicated everything. We'll also of course need to rewrite the various summary files. Do we do that in this PR? I think so right?

aganders3 commented:
At first glance I think yes.

Sorry but it's a holiday weekend here so I probably won't get a chance to look at this until Tuesday!

aganders3 commented:
Hm, on napari-hub that possibly could be resolved server-side by normalizing the name in the URL, but I'm not sure that will work with the static hub-lite. I'll have to look into if there's a simple mechanism for that, otherwise it should be possible to simply duplicate pages with normalized/non-normalized names. I'm not sure right now if that's a good idea or not.

This PR is at least a step towards _consistency_ in that regard, though.

DragaDoncila commented:
> I also learned "R[oute handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers)" will let you do this client-side as well with some different configuration. So you can indeed do something similar with a fully-static deployment (like hub-lite).

@aganders3 yep we discussed this at the Eurasia community meeting as well and we should be able to organize "redirects" from the non-normalized name to the normalized name and vice-versa. 

At community meeting we decided we should always store and key by normalized name, and all files e.g. manifest files are named by the normalized name, and the API should be accessed via its normalized name. I'm going to update this PR to reflect that. I think it should make things a lot tidier.

DragaDoncila commented:
Ok @aganders3 I think this is kinda ready for a review (leaving the conflicts for now and will resolve just before merge).

The current state is that **all** files are keyed by `normalized_name` and store the true `name` in their values. All individual plugin files including the individual manifests, pypi and conda files are named using the `normalized` plugin name. This was done by first clearing these folders and then running the indexing workflow from this branch to repopulate each folder. **Note** that the github folder has **not** been updated, because fetching github information is essentially broken (has been for a while it turns out). I think we should tackle that in a separate PR?

If you try to access a manifest via the API you can do so using the true plugin name OR the normalized name. Check out the vercel preview above to play with it.

**What does this mean for users of the API?**

- I've looked at `napari-plugin-manager` and it should not affect it. The plugin manager fetches `extended_summary` and keys into it, and we have not changed the value of any of the keys in the dictionary, just added a key. It does fetch the conda info, which now uses normalized names, but it was normalizing the package name anyway, so this should be a no-op.
- Based on what I can glean of `weather-report`, it should also be unaffected. It grabs the index to get the count of plugins, it grabs the `conda` info just to map to the plugin's conda name, and it grabs `extended_summary` but then normalizes the name anyway, so this should be a no-op. @Czaki maybe you can confirm.
- Hub only relies on the plugin index. Now, this index contains normalized names, but afaict this name will only be used to ping pypi for the plugin's info, and pypi takes normalized names with no issue.

aganders3 reviewed (approved):
This looks good to me! Maybe we should test the plugin manager against this manually by hard-coding the URL to the vercel preview? I can try to find time for that.

> Note that the github folder has not been updated, because fetching github information is essentially broken (has been for a while it turns out). I think we should tackle that in a separate PR?

I think that's okay - I'm not even familiar with what this is meant to be doing, so will need to look into it.

aganders3 commented:
After accounting for just differences in normalization (likely dupes on main?) I get this. I checked a few of these and they look like projects that have been _renamed_. Some also seem like they've been deleted from GitHub but not PyPI, maybe. I guess in general that does not cause the old versions to be removed, but I believe this PR starts with a clean slate. I need to think harder about what is desirable here.

```
>>> normalize - main
{'applet3',
'segmentree'}

>>> main - normalize
{'segtree',
'cryaleconcho',
'mkw-ropcop',
'appletree',
'napari-sim-simulator',
'motile-tracker',
'depalma-napari-omero',
'tracktour',
'ads-napari',
'napari-cryofibsem-imgpro',
'mmv-hitl4trk',
'napari-tomoslice',
'napari-brainways',
'morphospace'}
```

willingc commented on public/conda.json:
```diff
@@ -1,16 +1,33 @@
 {
+  "alveoleye": null,
```
Do we know why these are no longer sorted into the rest of the file? Was that by design?

willingc commented on scripts/reindex.py:
```diff
@@ -141,21 +145,30 @@ def conda_data_wrapper(args):
         print(f"{type(exc)}: {exc}", file=sys.stderr)
         active_pypi_versions = {}
 
+    # move the errors file to the top /public folder
+    desired_errors_path = PUBLIC / "errors.json"
+    current_error_path = Path(f"{PUBLIC}/manifest/errors.json")
+    if current_error_path.exists():
+        current_error_path.rename(desired_errors_path)
+    
```
I tend to agree with @aganders3. Was there a particular reason that you wanted to move the errors?

willingc commented on scripts/reindex.py:
```diff
@@ -141,21 +145,30 @@ def conda_data_wrapper(args):
+        name = info["name"]
+        pypi_versions = info["pypi_versions"]
+
+        mf_file = Path(f"{PUBLIC}/manifest/{normalized_name}.json")
+        if not mf_file.exists():
```
```suggestion
        manifest_file = Path(f"{PUBLIC}/manifest/{normalized_name}.json")
        if not manifest_file.exists():
```

I would reduce the cognitive load and be explicit in the variable name.

willingc commented on scripts/reindex.py:
```diff
@@ -176,10 +189,10 @@ def conda_data_wrapper(args):
 
     # sort things
-    PYPI_INDEX = sorted(PYPI_INDEX, key=lambda x: x["name"].lower())
+    PYPI_INDEX = sorted(PYPI_INDEX, key=lambda x: x["name"])
```
Yeah, it feels like we are duplicating sorting in too many places. I think some of these scripts could be combined.

willingc commented on scripts/utils.py:
```diff
@@ -0,0 +1,19 @@
+import re
```
I would recommend removing this file completely and use the `packaging` libraries convenience utilities for normalization of names. I believe we are already using packaging in the other scripts. https://packaging.pypa.io/en/stable/utils.html

DragaDoncila replied:
Done! Thanks for that tip!

willingc reviewed (approved):
@DragaDoncila While I'm not averse to merging as is, I think that there are a few improvements that can be made now, later, or never :wink:
- Split PR into scripts changes/pages changes from data changes (`public`). Not strictly necessary but a good practice in the future since it separates logic from applying the logic on the data.
- use packaging library's normalization utility functions
- reduce duplication across the scripts
- consider renaming script files to be more descriptive of their purpose in the workflow

DragaDoncila commented on scripts/reindex.py:
```diff
@@ -33,13 +35,15 @@
     author: str
     license: str
     home_page: str
+    project_url:  List[str]
```
Yep, just missed it in the other PR and mypy was complaining in VsCode. I've made #54 to track.

DragaDoncila commented:
I've reverted all data changes in this PR and addressed all the PR comments (I think...). I would say this is ready for merge, but we just need to figure out what the best way will be to handle the data updates.

We need to clear the `public/conda`, `public/manifest` and `public/pypi` directories and then run the update workflow, to make sure we start with a clean slate and no duplicate files as a result of the name normalization of files. 

The `public/github` directory is a different beast, idek what it's used for and the update step for it has been broken for a while. Opened #55 to track.

willingc reviewed (approved):
This looks great. Thank you for doing the detailed work. At some point after this lands, I may refactor the python files' print statements into log entries.

aganders3 commented:
> We need to clear the `public/conda`, `public/manifest` and `public/pypi` directories and then run the update workflow, to make sure we start with a clean slate and no duplicate files as a result of the name normalization of file names.

Do you want to clear these directories in this PR or a separate one?

willingc commented:
@aganders3 What's your preference: this PR or another? My gut says 2 PRs. One that does the clearing. And then this PR.

Thoughts?

aganders3 commented:
Yeah same - let's merge this as-is and open another PR that clears the directories and runs the workflow. I can probably open that!

willingc commented:
@DragaDoncila This looks like it is ready for you to merge.
